yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryProvider.scala

```diff
@@ -0,0 +1,1015 @@
+ * is synchronous and *not* cached.
+ * 5. the [[maybeCheckHealth()]] call performs a health check as the initial
+ * binding operation of this instance. This call invokes [[TimelineQueryClient.healthCheck()]]
+ * for better diagnostics on binding failures -particularly configuration problems.
+ * 6. Every REST call, synchronous or asynchronous, will invoke [[maybeCheckHealth()]] until
+ * the health check eventually succeeds.
+ * <p>
+ * If the timeline is  not enabled, the API calls used by the web UI
+ * downgrade gracefully (returning empty entries), rather than fail.
+ * 
+ *
+ * @param sparkConf configuration of the provider
+ */
+private[spark] class YarnHistoryProvider(sparkConf: SparkConf)
+  extends ApplicationHistoryProvider with Logging {
+
+  /**
+   * The configuration here is a YarnConfiguration built off the spark configuration
+   * supplied in the constructor; this operation ensures that `yarn-default.xml`
+   * and `yarn-site.xml` are pulled in. Options in the spark conf will override
+   * those in the -default and -site XML resources which are not marked as final.
+   */
+  private val yarnConf = {
+    new YarnConfiguration(SparkHadoopUtil.get.newConfiguration(sparkConf))
+  }
+
+  /**
+   * UI ACL option
+   */
+  private val uiAclsEnabled = sparkConf.getBoolean("spark.history.ui.acls.enable", false)
+
+  private val detailedInfo = sparkConf.getBoolean(YarnHistoryProvider.OPTION_DETAILED_INFO, false)
+  private val NOT_STARTED = "<Not Started>"
+
+  /* minimum interval between each check for event log updates */
+  private val refreshInterval = sparkConf.getLong(YarnHistoryProvider.OPTION_MIN_REFRESH_INTERVAL,
+    YarnHistoryProvider.DEFAULT_MIN_REFRESH_INTERVAL_SECONDS) * 1000
+
+  /**
+   * Window limit in milliseconds
+   */
+  private val windowLimitMs = sparkConf.getLong(YarnHistoryProvider.OPTION_WINDOW_LIMIT,
+    YarnHistoryProvider.DEFAULT_WINDOW_LIMIT) * 1000
+
+  /**
+   * Number of events to get
+   */
+  private val eventFetchLimit = sparkConf.getLong(YarnHistoryProvider.OPTION_EVENT_FETCH_LIMIT,
+    YarnHistoryProvider.DEFAULT_EVENT_FETCH_LIMIT)
+
+  private val eventFetchOption: Option[Long] = if (eventFetchLimit > 0) Some(eventFetchLimit) else None
+
+  /**
+   * Start time. Doesn't use the `now` call as tests can subclass that and
+   * it won't be valid until after the subclass has been constructed
+   */
+  val serviceStartTime = System.currentTimeMillis()
+
+  /**
+   * Timeline endpoint URI
+   */
+  protected val timelineEndpoint = createTimelineEndpoint()
+
+  /**
+   * The timeline query client which uses the `jersey`
+   * Jersey instance to talk to a timeline service running
+   * at [[timelineEndpoint]], and creates a timeline (write) client instance
+   * to handle token renewal
+   *
+   */
+  protected val timelineQueryClient = {
+    createTimelineQueryClient()
+  }
+
+
+  /**
+   * Override point: create the timeline endpoint
```

Comment:
What do you mean by "override point" in these comments? Are users meant to look at this? Or is it for testing? If the latter, could you use `@VisibleForTesting` instead?